src/stc/PlatWX.cpp

```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
-        doubleClickActionData = data;
+        Show(!event.IsIconized());
     }
 
+#elif !wxSTC_POPUP_IS_CUSTOM
 
-    void OnFocus(wxFocusEvent& event)
+    void wxSTCPopupWindow::OnFocus(wxFocusEvent& event)
     {
-        ActivateParent();
+        #if wxSTC_POPUP_IS_FRAME
+            ActivateParent();
+        #endif
+
         GetParent()->SetFocus();
         event.Skip();
     }
 
-    void OnSize(wxSizeEvent& event)
-    {
-        // resize the child
-        wxSize sz = GetClientSize();
-        lv->SetSize(sz);
-        // reset the column widths
-        lv->SetColumnWidth(0, IconWidth()+4);
-        lv->SetColumnWidth(1, sz.x - 2 - lv->GetColumnWidth(0) -
-                           wxSystemSettings::GetMetric(wxSYS_VSCROLL_X));
-        event.Skip();
-    }
+#endif // __WXOSX_COCOA__
 
-    void ActivateParent()
-    {
-        // Although we're a frame, we always want the parent to be active, so
-        // raise it whenever we get shown, focused, etc.
-        wxTopLevelWindow *frame = wxDynamicCast(
-            wxGetTopLevelParent(GetParent()), wxTopLevelWindow);
-        if (frame)
-            frame->Raise();
-    }
 
+//----------------------------------------------------------------------
+// Helper classes for ListBox
 
-    virtual void DoSetSize(int x, int y,
-                           int width, int height,
-                           int sizeFlags = wxSIZE_AUTO)
+// The class manages the colours, images, and other data needed for popup lists.
+class wxSTCListBoxVisualData
+{
+public:
+    wxSTCListBoxVisualData(int d);
+    virtual ~wxSTCListBoxVisualData();
+
+    // ListBoxImpl implementation
+    void SetDesiredVisibleRows(int d);
+    int  GetDesiredVisibleRows() const;
+    void RegisterImage(int type, const wxBitmap& bmp);
+    void RegisterImage(int, const char *);
+    void RegisterRGBAImage(int, int, int,const unsigned char *);
+    void ClearRegisteredImages();
+
+    // Image data
+    const wxBitmap* GetImage(int i) const;
+
+    // Colour data
+    void ComputeColours();
+    const wxColour& GetBorderColour() const;
+    void SetColours(const wxColour&, const wxColour&,
+                    const wxColour&,const wxColour&);
+    const wxColour& GetBgColour() const;
+    const wxColour& GetTextColour() const;
+    const wxColour& GetHighlightBgColour() const;
+    const wxColour& GetHighlightTextColour() const;
+
+    // ListCtrl Style
+    void UseListCtrlStyle(bool, const wxColour&, const wxColour&);
+    bool HasListCtrlAppearance() const;
+    const wxColour& GetCurrentBgColour() const;
+    const wxColour& GetCurrentTextColour() const;
+
+private:
+    WX_DECLARE_HASH_MAP(int, wxBitmap, wxIntegerHash, wxIntegerEqual, ImgList);
+
+    int      m_desiredVisibleRows;
+    ImgList  m_imgList;
+
+    wxColour m_borderColour;
+    wxColour m_bgColour;
+    wxColour m_textColour;
+    wxColour m_highlightBgColour;
+    wxColour m_highlightTextColour;
+    bool     m_bgColourIsSet;
```

Comment:
Why do we need these `xxxIsSet` variables? It seems strange to store the colour validity outside of it when we could just use `xxx.IsOk()` instead.

Reply:
The IsSet variables are tracking when the colors are set by the user.  An, admittedly exotic, example where they might be needed is in responding to system color changes.  When the `ComputeColours` function gets called, we need to recompute all the colors that have not been specifically set by the user.  At that point it won't be sufficient to check if the color is OK because all the colors will be OK due to a previous to the `ComputeColours` function.

I know there were other cases where those variables were needed, but that's the only one I remember right now.  I agree they're ugly and I tried to get rid of them, but I couldn't get everything to work without them.

Reply:
Sorry, I forgot to answer this one: I thought that perhaps we could avoid `xxxIsSet` by just checking if the colour is valid in `GetXXX()` and returning its default value (used in `ComputeColours()`) if it isn't.

If this doesn't work for some reason I'm missing, it's not a big deal to keep the `xxxIsSet`, of course, but I think they should be renamed to `xxxIsSetByUser` or, maybe, invert its meaning and call it `useDefaultXXX` because currently "is set" is rather confusing: it can be false even if the colour is, actually, set to a valid value.

Reply:
I've changed the names to m_useDefaultXXXColour.  That probably is a better name for the purpose those variables serve.